### ISSUE TITLE
AsSuperVisitor Declared_Primitive: type is not a box primitive.

### ISSUE THREAD

**Issue by wmdietlGC** [typetools/checker-framework#1407]
While trying to reproduce an issue similar to #1398 I ran into a different exception.
This is using #1387 with all included fixes there.

Take:

```java
import org.checkerframework.checker.nullness.qual.Nullable;

abstract class Crash7b {
  abstract <T> T foo(@Nullable T p1, @Nullable T p2);

  abstract <T extends Number> @Nullable T bar(int p1, T p2);

  int demo() {
    return foo(
        bar(5, 3),
        3);
  }
}
```

Running the Nullness Checker gives:

```
error: AsSuperVisitor Declared_Primitive: type is not a box primitive.
  Compilation unit: Crash7b.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:727)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:25)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Primitive(AsSuperVisitor.java:362)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Primitive(AsSuperVisitor.java:29)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:332)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:59)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:86)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:68)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:131)
  org.checkerframework.framework.type.DefaultTypeHierarchy.castedAsSuper(DefaultTypeHierarchy.java:1112)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Primitive(DefaultTypeHierarchy.java:611)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Primitive(DefaultTypeHierarchy.java:55)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:332)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:238)
  org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory$KeyForTypeHierarchy.isSubtype(KeyForAnnotatedTypeFactory.java:164)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:173)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:151)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:1814)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:1781)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitReturn(BaseTypeVisitor.java:1216)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitReturn(BaseTypeVisitor.java:160)
  com.sun.tools.javac.tree.JCTree$JCReturn.accept(JCTree.java:1390)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:68)
```

As I'm trying to go after a different bug, I didn't spend much time minimizing this test case. Maybe it can be simplified a bit more.

**Comment by wmdietlGC** [comment]
I think all nullness annotations can be removed to simplify this. Simplified test:

```java
abstract class Issue1407 {
  abstract <S> S foo(S p1, S p2);

  abstract <T extends Number> T bar(T p2);

  int demo() {
    return foo(bar(3), 3);
  }
}
```

**Comment by smillst** [comment]
Fixed in #1387.
